purity/gc/gc.go

```diff
@@ -463,72 +463,100 @@ func (c *collector) checkDirectories(dirs []*pb.Directory) {
 	}
 }
 
-func (c *collector) RemoveBlobs() error {
-	log.Notice("Determining blobs to remove...")
-	blobs := map[string][]*ppb.Blob{}
-	ars := map[string][]*ppb.Blob{}
+func (c *collector) RemoveActionResults() error {
+	log.Notice("Determining action results to remove...")
+	ars := []*ppb.Blob{}
 	numArs := 0
-	numBlobs := 0
 	var totalSize int64
 	for _, ar := range c.actionResults {
 		if c.shouldDelete(ar) {
 			log.Debug("Identified action result %s for deletion", ar.Hash)
-			key := ar.Hash[:2]
-			ars[key] = append(ars[key], &ppb.Blob{Hash: ar.Hash, SizeBytes: ar.SizeBytes})
-			delete(c.actionRFs, ar.Hash)
+			ars = append(ars, &ppb.Blob{Hash: ar.Hash, SizeBytes: ar.SizeBytes, CachePrefix: ar.CachePrefix})
 			totalSize += ar.SizeBytes
 			numArs++
+		} else {
+			c.liveActionResults[ar.Hash] = ar.SizeBytes
+		}
+	}
+	if c.dryRun {
+		log.Notice("Would delete %d action results, total size %s", numArs, humanize.Bytes(uint64(totalSize)))
+		return nil
+	}
+	log.Notice("Deleting %d action results, total size %s", numArs, humanize.Bytes(uint64(totalSize)))
+	ch := newProgressBar("Deleting blobs", 256)
+	defer func() {
+		close(ch)
+		time.Sleep(10 * time.Millisecond)
+	}()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+	for _, ar := range ars {
+		log.Debug("Removing action result %s", ar.Hash)
+		if _, err := c.gcclient.Delete(ctx, &ppb.DeleteRequest{
+			ActionResults: []*ppb.Blob{ar},
+			Hard:          true,
+		}); err != nil {
+			log.Debug("Failed to delete action result %s marking as live: %w", ar.Hash, err)
+			c.liveActionResults[ar.Hash] = ar.SizeBytes
+		} else {
+			log.Debug("Deleted action result: %s", ar.Hash)
+			delete(c.actionRFs, ar.Hash)
 		}
+		ch <- 1
 	}
-	for hash, size := range c.allBlobs {
+	return nil
+}
+
+func (c *collector) RemoveBlobs() error {
+	log.Notice("Determining blobs to remove...")
+	blobs := []*ppb.Blob{}
```

Comment:
nit: this is almost always going to close to the size of `referencedBlobs`, right? If it is, it might be a good idea to initialise the slice to `make([]*ppb.Blob, 0, len(c.referencedBlobs))`

Reply:
c.referencedBlobs don't get deleted. I have set the capacity to c.allBlobs - c.referencedBlobs.

Reply:
Ah yeah, good point. Although I think you forgot to push that change